core/src/main/java/org/infinispan/interceptors/impl/CacheMgmtInterceptor.java

```diff
@@ -70,63 +70,62 @@ public void start() {
 
    @Override
    public CompletableFuture<Void> visitEvictCommand(InvocationContext ctx, EvictCommand command) throws Throwable {
-      Object returnValue = ctx.forkInvocationSync(command);
-      if (getStatisticsEnabled(command))
-         evictions.increment();
+      if (!getStatisticsEnabled(command))
+         return ctx.continueInvocation();
 
-      return ctx.shortCircuit(returnValue);
+      return ctx.onReturn((ctx1, command1, rv, throwable) -> {
+         evictions.increment();
+         return null;
+      });
    }
 
    @Override
    public final CompletableFuture<Void> visitGetKeyValueCommand(InvocationContext ctx, GetKeyValueCommand command) throws Throwable {
       return visitDataReadCommand(ctx, command);
    }
+
    @Override
    public final CompletableFuture<Void> visitGetCacheEntryCommand(InvocationContext ctx, GetCacheEntryCommand command) throws Throwable {
       return visitDataReadCommand(ctx, command);
    }
+
    private CompletableFuture<Void> visitDataReadCommand(InvocationContext ctx, AbstractDataCommand command) throws Throwable {
-      long start = 0;
       boolean statisticsEnabled = getStatisticsEnabled(command);
-      if (statisticsEnabled)
-         start = timeService.time();
-
-      Object retval = ctx.forkInvocationSync(command);
+      if (!statisticsEnabled || !ctx.isOriginLocal())
+         return ctx.continueInvocation();
 
-      if (statisticsEnabled && ctx.isOriginLocal()) {
+      long start = timeService.time();
+      return ctx.onReturn((ctx1, command1, rv, throwable) -> {
          long intervalMilliseconds = timeService.timeDuration(start, TimeUnit.MILLISECONDS);
```

Comment:
isn't it possible the ```ReturnHandler``` to be invoke on a different thread? I'm wondering if our stats would be less reliable now...